core/src/main/scala/cats/MonadError.scala

```diff
@@ -13,6 +13,23 @@ trait MonadError[F[_], E] extends ApplicativeError[F, E] with Monad[F] {
   def ensure[A](fa: F[A])(error: => E)(predicate: A => Boolean): F[A] =
     flatMap(fa)(a => if (predicate(a)) pure(a) else raiseError(error))
 
+  /**
+   * Sequences the specified finalizer ensuring evaluation regardless of
+   * whether or not the target `F[A]` raises an error.
+   *
+   * If `raiseError` is analogous to `throw` and `handleErrorWith` is analogous to
+   * `catch`, then `guarantee` is analogous to `finally`.  JVM exception semantics
+   * are mirrored with respect to error raised within the finalizer (i.e. errors
+   * raised within the finalizer will shade any errors raised by the primary action).
+   *
+   * @see [[raiseError]]
+   * @see [[handleErrorWith]]
+   */
+  def guarantee[A](fa: F[A], finalizer: F[Unit]): F[A] = {
```

Comment:
I need to see tests here specified in terms of the "bracket problem" (see http://hackage.haskell.org/package/layers-0.1/docs/Documentation-Layers-Overview.html#g:9 and the next couple sections).

I am not convinced this belongs on MonadError. As we've discussed previously, a monad layer being able to return an error has nothing to do with the *rest* of the effects in the stack being able to halt the computation, which is what `guarantee` exists to guarantee semantics for.